## **01_data_clean**

### Objectives

* Fetch data from Kaggle and save as raw data
* Clean the data

### Inputs

* https://www.kaggle.com/datasets/rabieelkharoua/predict-pet-adoption-status-dataset/data
* https://doi.org/10.34740/kaggle/ds/5242440
* Save as data_raw.csv

### Outputs

* Save cleaned data as data_clean.csv

### Additional Comments

* 



In [1]:
# import required libraries
import os
import pandas as pd
import numpy as np

We need to change the working directory from its current folder to its parent folder
* We access the current directory with os.getcwd()

In [2]:
# get the current working directory
current_dir = os.getcwd()
current_dir

'c:\\Users\\beth_\\Documents\\vscode-projects\\Animal-Adoption-Capstone\\jupyter_notebooks'

We want to make the parent of the current directory the new current directory
* os.path.dirname() gets the parent directory
* os.chir() defines the new current directory

In [3]:
os.chdir(os.path.dirname(current_dir))
print("You set a new current directory")

You set a new current directory


Confirm the new current directory

In [4]:
current_dir = os.getcwd()
current_dir

'c:\\Users\\beth_\\Documents\\vscode-projects\\Animal-Adoption-Capstone'

## Saving the data to data_raw.csv

In [18]:
# load the raw data
# use of Copilot to access the correct file path
df = pd.read_csv(os.path.join(current_dir, 'data', 'data_raw.csv'))
df.head(10)

PetID PetType             Breed  AgeMonths   Color    Size   WeightKg  \
0    500    Bird          Parakeet        131  Orange   Large   5.039768   
1    501  Rabbit            Rabbit         73   White   Large  16.086727   
2    502     Dog  Golden Retriever        136  Orange  Medium   2.076286   
3    503    Bird          Parakeet         97   White   Small   3.339423   
4    504  Rabbit            Rabbit        123    Gray   Large  20.498100   
5    505     Dog          Labrador         70   Brown   Large  20.986261   
6    506    Bird          Parakeet        169   Brown   Small  10.902613   
7    507     Cat           Siamese         13  Orange   Large   7.252683   
8    508    Bird          Parakeet         49   Brown  Medium  24.597598   
9    509    Bird          Parakeet         60    Gray   Large   7.295994   

   Vaccinated  HealthCondition  TimeInShelterDays  AdoptionFee  PreviousOwner  \
0           1                0                 27          140              0   
1           0                0                  8          235              0   
2           0                0                 85          385              0   
3           0                0                 61          217              1   
4           0                0                 28           14              1   
5           0                0                 87          301              1   
6           1                0                 70          440              1   
7           1                0                  3          137              0   
8           1                1                 69          405              0   
9           0                0                 73          231              1   

   AdoptionLikelihood  
0                   0  
1                   0  
2                   0  
3                   0  
4                   0  
5                   0  
6                   0  
7                   1  
8                   0  
9                   0

## Initial data exploration and cleaning

- missing values
- duplicates
- outliers
- convert data types
- data ethics
- rename/remove columns

In [6]:
# start to find out about the data
df.shape

(2007, 13)

In [7]:
# get a feel for the data types and non-null counts
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2007 entries, 0 to 2006
Data columns (total 13 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   PetID               2007 non-null   int64  
 1   PetType             2007 non-null   object 
 2   Breed               2007 non-null   object 
 3   AgeMonths           2007 non-null   int64  
 4   Color               2007 non-null   object 
 5   Size                2007 non-null   object 
 6   WeightKg            2007 non-null   float64
 7   Vaccinated          2007 non-null   int64  
 8   HealthCondition     2007 non-null   int64  
 9   TimeInShelterDays   2007 non-null   int64  
 10  AdoptionFee         2007 non-null   int64  
 11  PreviousOwner       2007 non-null   int64  
 12  AdoptionLikelihood  2007 non-null   int64  
dtypes: float64(1), int64(8), object(4)
memory usage: 204.0+ KB


In [ ]:
# in the data the index does not need to be reset as it is in order
# but if needed, uncomment the lines below to reset the index
# df.reset_index(drop=True, inplace=True)
# df.head()

In [8]:
# see if there are any missing values
df.isna().sum()

PetID                 0
PetType               0
Breed                 0
AgeMonths             0
Color                 0
Size                  0
WeightKg              0
Vaccinated            0
HealthCondition       0
TimeInShelterDays     0
AdoptionFee           0
PreviousOwner         0
AdoptionLikelihood    0
dtype: int64

There are no missing values to deal with.
If there were missing values, depending on the type of missing values, I would consider either replacing with the median or modal value, or inserting the value in the row above/below. The data set is not that large and it would be a shame to delete whole rows and lose data. 

In [9]:
# check for duplicates
dupes = df.duplicated().sum()
print("Duplicate rows: ", dupes) 
duplicates_all = df[df.duplicated(keep=False)]


Duplicate rows:  0


There are no duplicated rows, if there were, the code is shown below:

In [10]:
# drop the duplicate row and modify the dataframe in place
# df.drop_duplicates(df, keep='first', inplace=True)
# df.shape

In [11]:
# see what the summary statistics look like
# looking at the minimum and maximum values to see if there are any obvious anomalies
df.describe()

PetID    AgeMonths     WeightKg   Vaccinated  HealthCondition  \
count  2007.000000  2007.000000  2007.000000  2007.000000      2007.000000   
mean   1503.000000    92.279522    15.705776     0.701046         0.196313   
std     579.515315    52.148363     8.327749     0.457914         0.397307   
min     500.000000     1.000000     1.018198     0.000000         0.000000   
25%    1001.500000    48.000000     8.730396     0.000000         0.000000   
50%    1503.000000    94.000000    15.925416     1.000000         0.000000   
75%    2004.500000   138.000000    22.737180     1.000000         0.000000   
max    2506.000000   179.000000    29.995628     1.000000         1.000000   

       TimeInShelterDays  AdoptionFee  PreviousOwner  AdoptionLikelihood  
count        2007.000000  2007.000000    2007.000000         2007.000000  
mean           43.974091   249.142003       0.301943            0.328351  
std            25.740253   142.887040       0.459215            0.469730  
min             1.000000     0.000000       0.000000            0.000000  
25%            21.000000   127.000000       0.000000            0.000000  
50%            45.000000   242.000000       0.000000            0.000000  
75%            66.000000   375.000000       1.000000            1.000000  
max            89.000000   499.000000       1.000000            1.000000

There are no negative values, if there were I would replace with 0.0.
The maximum values look sensible: 179 months is 14.92 years; 30kg for a large dog is normal.

In [17]:
# declare categorical types where they are object
# this will help with memory usage and clarity
category_cols = ["PetType", "Breed", "Colour","Size"]
for cat in category_cols:
    df[cat] = df[cat].astype("category")

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2007 entries, 0 to 2006
Data columns (total 13 columns):
 #   Column              Non-Null Count  Dtype   
---  ------              --------------  -----   
 0   PetType             2007 non-null   category
 1   Breed               2007 non-null   category
 2   AgeMonths           2007 non-null   int64   
 3   Colour              2007 non-null   category
 4   Size                2007 non-null   category
 5   WeightKg            2007 non-null   float64 
 6   Vaccinated          2007 non-null   int64   
 7   HealthCondition     2007 non-null   int64   
 8   TimeInShelterDays   2007 non-null   int64   
 9   AdoptionFee         2007 non-null   int64   
 10  PreviousOwner       2007 non-null   int64   
 11  AdoptionLikelihood  2007 non-null   int64   
 12  AgeInYears          2007 non-null   float64 
dtypes: category(4), float64(2), int64(7)
memory usage: 150.0 KB


In [12]:
# remove PetID column as it is not needed for analysis
df.drop(columns=['PetID'], inplace=True)
df.head()


PetType             Breed  AgeMonths   Color    Size   WeightKg  Vaccinated  \
0    Bird          Parakeet        131  Orange   Large   5.039768           1   
1  Rabbit            Rabbit         73   White   Large  16.086727           0   
2     Dog  Golden Retriever        136  Orange  Medium   2.076286           0   
3    Bird          Parakeet         97   White   Small   3.339423           0   
4  Rabbit            Rabbit        123    Gray   Large  20.498100           0   

   HealthCondition  TimeInShelterDays  AdoptionFee  PreviousOwner  \
0                0                 27          140              0   
1                0                  8          235              0   
2                0                 85          385              0   
3                0                 61          217              1   
4                0                 28           14              1   

   AdoptionLikelihood  
0                   0  
1                   0  
2                   0  
3                   0  
4                   0

PetID is not required for analysis. However, it is also best to remove this identifier to fully anonymise the data. If this were a real world data set any identifiers of specific animals or details of the owners would need to be anonymised using a boolean mask as shown in the code below. 

In [13]:
# get unique PetID
# unique_pet_id = df['PetID'].unique()
# map new IDs to replace the old ones
# pet_id_map = {num: f'Pet-{i+1}' for i, num in enumerate(unique_pet_id)}
# make a new column with the anonymous IDs
# df['anon_petID'] = df['PetID'].map(pet_id_map)

In [14]:
# rename the Color column to Colour
df.rename(columns={'Color': 'Colour'}, inplace=True)
df.head()

PetType             Breed  AgeMonths  Colour    Size   WeightKg  Vaccinated  \
0    Bird          Parakeet        131  Orange   Large   5.039768           1   
1  Rabbit            Rabbit         73   White   Large  16.086727           0   
2     Dog  Golden Retriever        136  Orange  Medium   2.076286           0   
3    Bird          Parakeet         97   White   Small   3.339423           0   
4  Rabbit            Rabbit        123    Gray   Large  20.498100           0   

   HealthCondition  TimeInShelterDays  AdoptionFee  PreviousOwner  \
0                0                 27          140              0   
1                0                  8          235              0   
2                0                 85          385              0   
3                0                 61          217              1   
4                0                 28           14              1   

   AdoptionLikelihood  
0                   0  
1                   0  
2                   0  
3                   0  
4                   0

In [15]:
# create a new column to show the age of the pet in years
df['AgeInYears'] = df['AgeMonths']/12
df.head()

PetType             Breed  AgeMonths  Colour    Size   WeightKg  Vaccinated  \
0    Bird          Parakeet        131  Orange   Large   5.039768           1   
1  Rabbit            Rabbit         73   White   Large  16.086727           0   
2     Dog  Golden Retriever        136  Orange  Medium   2.076286           0   
3    Bird          Parakeet         97   White   Small   3.339423           0   
4  Rabbit            Rabbit        123    Gray   Large  20.498100           0   

   HealthCondition  TimeInShelterDays  AdoptionFee  PreviousOwner  \
0                0                 27          140              0   
1                0                  8          235              0   
2                0                 85          385              0   
3                0                 61          217              1   
4                0                 28           14              1   

   AdoptionLikelihood  AgeInYears  
0                   0   10.916667  
1                   0    6.083333  
2                   0   11.333333  
3                   0    8.083333  
4                   0   10.250000

## Save the cleaned data to a new file data_clean.csv

In [16]:
# save the cleaned data to a new csv file
df.to_csv(os.path.join(current_dir, 'data', 'data_clean.csv'), index=False)


# Conclusions and next steps
- The data is now saved to data_clean.csv
- Next is EDA - exploratory data analysis
- Please go to notebook 02_eda_visuals.ipynb